errbot/backends/irc.py

```diff
@@ -481,8 +539,61 @@ def on_join(self, connection, event):
         # We need to wait to endofnames message.
         room_name = event.target
         with self._recently_joined_lock:
+            if room_name not in self.bot._rooms:
+                self.bot._rooms[room_name] = IRCRoom(room_name, self.bot)
```

Comment:
`self.bot._rooms` needs to be protected with a thead lock.

I would suggest renaming `self._recently_joined_lock` to `self._room_lock` and just reusing that (it wouldn't really add any value to use two separate locks for this here)